File: src/myxo/rules_validator.py
"""Validator for .myxo/rules.md brevity guidelines."""

MAX_LINES = 50
MAX_LINE_LENGTH = 120


def validate_rules(content: str) -> list[str]:
    """Validate rules.md content and return a list of violation messages.

    Rules:
    - Each line must start with ``- `` (bullet), ``#`` (header), or be blank.
    - Maximum 50 lines total.
    - Each line must be at most 120 characters.
    """
    if not content.strip():
        return []

    errors: list[str] = []
    lines = content.strip().split("\n")

    if len(lines) > MAX_LINES:
        errors.append(
            f"rules.md exceeds {MAX_LINES} lines (found {len(lines)})"
        )

    for i, line in enumerate(lines, start=1):
        stripped = line.strip()

        # Allow blank lines and header lines
        if stripped == "" or stripped.startswith("#"):
            continue

        # Must be a bullet line
        if not stripped.startswith("- "):
            errors.append(
                f"Line {i}: must start with '- ' (bullet point), got: {stripped[:40]}"
            )

        if len(line) > MAX_LINE_LENGTH:
            errors.append(
                f"Line {i}: exceeds {MAX_LINE_LENGTH} characters (found {len(line)})"
            )

    return errors

File: crates/myxo-core/src/rules_validator.rs
const MAX_LINES: usize = 50;
const MAX_LINE_LENGTH: usize = 120;

pub fn validate_rules(content: &str) -> Vec<String> {
    if content.trim().is_empty() {
        return vec![];
    }

    let mut errors = Vec::new();
    let mut lines: Vec<&str> = content.split('\n').collect();

    // Remove trailing empty line caused by final newline
    if lines.last() == Some(&"") {
        lines.pop();
    }

    if lines.len() > MAX_LINES {
        errors.push(format!(
            "rules.md exceeds {MAX_LINES} lines (found {})",
            lines.len()
        ));
    }

    for (i, line) in lines.iter().enumerate() {
        let line_num = i + 1;

        if line.len() > MAX_LINE_LENGTH {
            errors.push(format!(
                "Line {line_num}: exceeds {MAX_LINE_LENGTH} characters (found {})",
                line.len()
            ));
        }

        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        if !line.starts_with("- ") {
            let preview: String = line.chars().take(40).collect();
            errors.push(format!(
                "Line {line_num}: must start with '- ' (bullet point), got: {preview}"
            ));
        }
    }

    errors
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_rules_no_errors() {
        let content = "# Rules\n\n- Rule one\n- Rule two\n";
        let errors = validate_rules(content);
        assert!(errors.is_empty());
    }

    #[test]
    fn empty_content_no_errors() {
        let errors = validate_rules("");
        assert!(errors.is_empty());
    }

    #[test]
    fn invalid_line_format() {
        let content = "# Rules\n\nNot a bullet line\n";
        let errors = validate_rules(content);
        assert!(!errors.is_empty());
        assert!(errors[0].contains("must start with '- '"));
    }

    #[test]
    fn exceeds_max_lines() {
        let content = (0..51)
            .map(|i| format!("- Rule {i}"))
            .collect::<Vec<_>>()
            .join("\n");
        let errors = validate_rules(&content);
        assert!(errors.iter().any(|e| e.contains("exceeds 50 lines")));
    }

    #[test]
    fn exceeds_max_line_length() {
        let long_line = format!("- {}", "a".repeat(120));
        let errors = validate_rules(&long_line);
        assert!(errors.iter().any(|e| e.contains("exceeds 120 characters")));
    }
}
